drivers/afe/ad413x/ad413x.c

```diff
@@ -0,0 +1,816 @@
+*******************************************************************************/
+int32_t ad413x_continuous_read(struct ad413x_dev *dev, int32_t *buffer,
+			       int32_t ch_nb, int32_t sample_nb)
+{
+	int32_t ret;
+
+	struct ad413x_callback_ctx ctx = {
+		.dev = dev,
+		.buffer = buffer,
+		.buffer_size = ch_nb * sample_nb,
+	};
+
+	struct callback_desc irq_callback = {
+		.callback = &irq_adc_cont_read,
+		.ctx = &ctx
+	};
+
+	ret = irq_trigger_level_set(dev->irq_desc, dev->rdy_pin, IRQ_LEVEL_LOW);
+	if (ret)
+		return ret;
+	ret = irq_register_callback(dev->irq_desc, dev->rdy_pin, &irq_callback);
+	if (ret)
+		return ret;
+
+	ret = ad413x_spi_int_reg_write_msk(dev,
+					   AD413X_REG_ADC_CTRL,
+					   AD413X_ADC_CONT_READ,
+					   AD413X_ADC_CONT_READ);
+	if (ret)
+		return ret;
+
+	ret = ad413x_set_adc_mode(dev, AD413X_CONTINOUS_CONV_MODE);
+	if (ret)
+		return ret;
+
+	// make sure adc is fully initialized before irq enabling
+	mdelay(1U);
+
+	ret = irq_enable(dev->irq_desc, dev->rdy_pin);
+	if (ret)
+		return ret;
+
+	while(ctx.buffer_size != 0U) ;
+
+	return ad413x_spi_int_reg_write_msk(dev,
+					    AD413X_REG_ADC_CTRL,
+					    0x00,
+					    AD413X_ADC_CONT_READ);
+}
+
+/***************************************************************************//**
+ * Store adc channel presets.
+ *
+ * @param dev        - The device structure.
+ * @param preset 	 - The structure to be saved as preset.
+ * @param preset_nb	 - Preset's number.
+ *
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t ad413x_preset_store(struct ad413x_dev *dev, struct ad413x_preset preset,
+			    uint8_t preset_nb)
+{
+	int32_t ret;
+
+	ret = ad413x_set_gain(dev, preset.gain, preset_nb);
+	if (ret)
+		return ret;
+
+	ret = ad413x_set_ref(dev, preset.ref_sel, preset_nb);
+	if (ret)
+		return ret;
+
+	return ad413x_set_filter(dev, preset.filter, preset_nb);
+}
+
+/***************************************************************************//**
+ * Initialize the device.
+ *
+ * @param device     - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ *
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t ad413x_init(struct ad413x_dev **device,
+		    struct ad413x_init_param init_param)
+{
+	struct ad413x_dev *dev;
+	uint32_t reg_data;
+	int32_t ret;
+	int32_t i;
+
+	dev = (struct ad413x_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	dev->chip_id = init_param.chip_id;
+	dev->int_ref = init_param.int_ref;
+
+	/* SPI */
+	ret = spi_init(&dev->spi_dev, init_param.spi_init);
+	if (ret)
+		goto err_dev;
+
+	/* Device Settings */
+	ret = ad413x_spi_do_soft_reset(dev);
+	if (ret)
+		goto err_spi;
+
+	/* Change SPI to 4 wire*/
+	ret = ad413x_spi_int_reg_write_msk(dev,
```

Comment:
it think we should expose a device parameter `spi3wire` as boolean, and give the user the option to selecte between 3-wire configuration and 4-wire configuration